### ISSUE TITLE
P3 plateau Phase 3: long-horizon (500 iter) test of L1-anchored configs

### ISSUE THREAD

**Issue by rjwalters** [rjwalters/bucket-brigade#183]
## Context

#174 ran the 3-lever PPO ablation on scenario `default`, λ_red=0, 5 seeds × 50 iters (PR #182). Result: no condition crosses 320, none even crosses the 308 random baseline. L1 (return normalization) crushes the value-loss-dominance ratio from 6e+06 to 2e+01 and prevents entropy collapse, but reward never moves.

The only positive signal in the Phase 2 data is the slope in iters 25–50:
- L1_norm:    **+0.173 reward/iter**
- L1+L2:      +0.125
- baseline:   +0.090

Linear extrapolation to iter 500 puts L1_norm at ~377 — above the 320 bar. Slopes typically decay, but the trajectory shape is non-flat in the right direction. **50 iters may simply be too short to see PPO escape this basin.**

This issue is the long-horizon test of that hypothesis. If it passes, the verdict from #174 flips ("the levers do work, just need horizon"). If it fails, the answer is *not* in PPO hyperparameter space and we move to one of: reward-attribution rework (#135), different advantage estimator, different algorithm.

## What to run

| axis | values |
|---|---|
| scenario | `default`, `chain_reaction` |
| λ_red | 0 |
| condition | `L1_norm` (T,0.5,0.01), `L1L2` (T,0.05,0.01), `L1L3` (T,0.5,0.1) |
| seeds | 42..46 (5) |
| num_iterations | **500** |
| rollout_steps | 2048 |
| device | cpu |

= 2 × 3 × 5 = **30 cells × ~10× per-cell wall clock**.

Mac Studio CPU estimate: ~50 hours single-threaded, ~6–8 hours at 8-way process parallelism (cells are seed-independent — easy to shard).
Alc-cluster (RTX 4090, 32 cores) estimate: ~30 min at 16-way (env is CPU-bound; GPU buys little per-cell but parallelism is the win).

Recommend running on **alc-2** to keep Mac Studio free. Cells are seed-independent so sharding is just splitting the seeds list across 3 shells.

## Report

`research_notebook/<date>_p3_plateau_phase3.md` with:

- Final-iter reward CI per (condition, scenario) cell, vs random baseline.
- Reward trajectory plot (per condition, per scenario) overlaid with the Phase 2 trajectories from `results_174_ablation/reward_trajectories.png` for visual continuity.
- Updated `gradient_ratios` plot showing whether the ratio stays balanced over 10× longer training.
- Whether **any** (condition, scenario) cell crosses CI lower bound > 320 (the #174 bar).
- Whether `chain_reaction` behaves differently from `default` (P3's behavioral signal in the May 14 result lived on `chain_reaction`).

## Acceptance criteria

- At least one (condition, scenario) cell has final-iter reward CI lower bound > 320.
- **If yes**: open follow-up PR setting the winning combination as `CellConfig` defaults (this completes the work that #174's acceptance criterion deferred).
- **If no**: write the negative result; the next issue moves out of PPO-hyperparameter space (re-examine #135 reward attribution, or trial a different algorithm).

## What this unblocks

- **P3 v2 sweep** (3 × 4 × 20 = 240 cells overnight) on the fixed env + fixed plateau. Currently still blocked: re-running with #174's hyperparameters but no learning produces clean gradients on broken agents.
- **#176** validation of state-summary CMI conditioner — needs policies that actually learn before the conditioner can be meaningfully tested.

## Related

- #174 (Phase 2 ablation, PR #182 closes)
- #145 (Phase 1 diagnostic — PPO doesn't learn)
- #151 (Phase 1 plots)
- #168 (state-summary CMI conditioner)
- #135 (reward-attribution Python vs Rust core — out-of-PPO-hyperparameter-space candidate)

**Comment by rjwalters** [comment]
## Curator Enhancement

### Problem summary

Phase 2 (#174, PR #182) ran the 3-lever PPO ablation on `default` at λ_red=0, 5 seeds × 50 iters. **No condition crossed the 320 acceptance bar; none even crossed the 308 random baseline.** The only positive signal was a non-flat slope in iters 25-50: L1_norm at +0.173 reward/iter, L1+L2 at +0.125, baseline at +0.090. Linear (and likely optimistic) extrapolation puts L1_norm at ~377 by iter 500 — over the bar. This issue tests whether 50 iters was just too short for PPO to escape the basin, by running the 3 L1-anchored configs at 10× the horizon on **two** scenarios (`default` + `chain_reaction`, since the latter is where the May 14 P3 behavioral signal lived).

### Concrete CLI invocation

Cells are seed-independent and per-lambda-independent → trivial to shard. The three condition launches (assumed cwd = repo root in a `uv` env, run after `git pull` on the target host):

```bash
# Condition L1_norm: (normalize_returns=T, value_coef=0.5, entropy_coef=0.01)
uv run python experiments/p3_specialization/run_sweep.py \
  --output-root experiments/p3_specialization/runs_183_L1_norm \
  --scenarios default chain_reaction \
  --lambdas 0 \
  --seeds 42 43 44 45 46 \
  --num-iterations 500 \
  --rollout-steps 2048 \
  --device cpu \
  --normalize-returns

# Condition L1L2: (normalize_returns=T, value_coef=0.05, entropy_coef=0.01)
uv run python experiments/p3_specialization/run_sweep.py \
  --output-root experiments/p3_specialization/runs_183_L1L2 \
  --scenarios default chain_reaction \
  --lambdas 0 \
  --seeds 42 43 44 45 46 \
  --num-iterations 500 \
  --rollout-steps 2048 \
  --device cpu \
  --normalize-returns \
  --value-coef 0.05

# Condition L1L3: (normalize_returns=T, value_coef=0.5, entropy_coef=0.1)
uv run python experiments/p3_specialization/run_sweep.py \
  --output-root experiments/p3_specialization/runs_183_L1L3 \
  --scenarios default chain_reaction \
  --lambdas 0 \
  --seeds 42 43 44 45 46 \
  --num-iterations 500 \
  --rollout-steps 2048 \
  --device cpu \
  --normalize-returns \
  --entropy-coef 0.1
```

Each invocation: 2 scenarios × 1 λ × 5 seeds = 10 cells. Three invocations × 10 cells = **30 cells total**, matching the issue body's count.

Run-host guidance: per `CLAUDE.md` the GPU-friendly host is `rwalters-sandbox-2`, but the env here is CPU-bound (PPO + small networks; `--device cpu` is set). Either `rwalters-sandbox-2` or the alc cluster works; the win is per-cell parallelism, not per-cell GPU. Confirm the issue body's "~30 min at 16-way parallelism on alc-2" wall-clock estimate empirically on the first shard before launching the rest. (Per-cell wall-clock scales ~linearly with `--num-iterations`; 50 iters ≈ 63s on Mac Studio CPU per #174, so 500 iters ≈ 10-12 min/cell single-threaded, ~30 min × 30 cells / 16 workers ≈ 56 cell-min ≈ feasible if the alc nodes are fully free.)

Sharding strategy if needed: split `--seeds 42 43 44 45 46` across three shells (e.g. one shell each for seeds {42,43}, {44}, {45,46}) or split scenarios. The directory layout (`runs_183_<COND>/<scenario>/lambda_0e+00/seed_<N>/`) is collision-free across shards.

### Affected Files

- `experiments/p3_specialization/run_sweep.py:104-148` — **REFERENCE ONLY**. CLI entrypoint already exposes `--normalize-returns`, `--value-coef`, `--entropy-coef`, `--num-iterations`, `--scenarios`, `--lambdas`, `--seeds`, `--device`. No plumbing change needed.
- `experiments/p3_specialization/train.py:57-76` — **REFERENCE ONLY**. `CellConfig` defaults (`value_coef=0.5`, `entropy_coef=0.01`, `normalize_returns=False`) are what each invocation overrides. Do **not** edit defaults in this PR.
- `experiments/p3_specialization/train.py:459-462` — **REFERENCE ONLY**. Where `CellConfig` fields are passed through to the trainer.
- `bucket_brigade/training/normalizers.py` — **REFERENCE ONLY**. Welford running-stats normalizer landed in #159 (commit 376366dc); activated by `--normalize-returns`.
- `experiments/p3_specialization/results_174_ablation/reward_trajectories.png` — **INPUT**. The Phase 2 trajectories to overlay for visual continuity in the new plot.
- `experiments/p3_specialization/runs_183_{L1_norm,L1L2,L1L3}/` — **NEW**. Per-shard output directories. Gitignored per existing `runs/` pattern.
- `experiments/p3_specialization/analyze_183.py` — **NEW (likely)**. Aggregation + plotting analogous to `analyze_174.py` from PR #182. Produces `results_183/{summary.json,summary.md,reward_trajectories.png,gradient_ratios.png}`. Mirror the `analyze_174.py` layout for consistency.
- `research_notebook/2026-MM-DD_p3_plateau_phase3.md` — **NEW**. Report deliverable with the sections enumerated in the issue body.

### Test Plan

- [ ] **Smoke check first**: launch one cell of L1_norm (`--seeds 42 --num-iterations 20`) and confirm `metrics.json` is produced and the directory layout matches. Aborts cheap if the longer horizon hits an unexpected NaN / OOM.
- [ ] **Sweep completion**: 30/30 cells write `metrics.json`. Re-runnable shards via `--skip-existing`.
- [ ] **Primary AC**: at least one (condition, scenario) cell has final-iter (iter 499) reward CI lower bound > 320, bootstrap 95% CI over 5 seeds.
- [ ] **Secondary check**: confirm or refute that the iter 25-50 slope is a real trajectory feature, not a fluctuation. Acceptable evidence: the iter 250-500 slope retains the same sign as iter 25-50, even if the magnitude has decayed (mid-training slopes typically halve every ~2× horizon under PPO).
- [ ] **Scenario diff check**: report whether `chain_reaction` is materially different from `default` at iter 499. P3's behavioral signal in the May 14 result lived on `chain_reaction`; if 500 iters reveals a regime difference, that's reportable independent of whether the 320 bar is crossed.
- [ ] **Gradient sanity**: updated `gradient_ratios.png` confirms the value-loss-dominance ratio remains in the L1-corrected 10¹-10² range (not regressing toward 10⁶) across the full 500 iters.
- [ ] **Reward-trajectory plot**: overlay Phase 3 trajectories on top of Phase 2 trajectories from `results_174_ablation/reward_trajectories.png` (x-axis = iter, 0-500).

**Positive outcome path**: file follow-up PR setting the winning (`normalize_returns`, `value_coef`, `entropy_coef`) tuple as `CellConfig` defaults. This completes the work that #174's acceptance criterion deferred. Unblocks #176 (which was waiting on #174's defaults change; see Dependency note below) and the P3 v2 sweep.

**Negative outcome path**: write the negative result. Next experiment moves out of PPO-hyperparameter space — re-examine #135 (reward-attribution Python vs Rust core) or trial a different algorithm. #176 then needs reframing because "policies that actually learn" is no longer available at all under the current ablation budget.

### Scope guards

- **Do not** change `CellConfig` defaults in this PR. A defaults change is gated on this experiment producing CI lower bound > 320, and even then it belongs in a separate follow-up PR (mirrors #174's structure: ablation in one PR, defaults change in another).
- **Do not** run the full pre-registered P3 v2 sweep (3 × 4 × 20 = 240 cells) here. That is a separate downstream issue gated on this result.
- **Do not** modify `experiments/p3_specialization/train.py` or `run_sweep.py` in scope of this issue. The CLI already supports every required toggle.
- **Do not** add new λ_red values; this issue is explicitly λ=0 to isolate the learning-vs-coordination question. λ>0 cells are #176's territory.
- **Do not** broaden the seed count beyond 42..46. The 5-seed × 500-iter × 2-scenario × 3-condition cube is already at the upper end of "feasible on a shared compute host."
- If wall-clock per cell exceeds ~30 min during the smoke shard, **stop and re-curate** rather than over-running the cluster budget.

### Dependency note — relationship to #176

**This issue is now the de facto blocker for #176.** Before PR #182 merged, #176's documented blocker was "#174 must identify a winning lever configuration." #174 returned a negative result: no lever (alone or paired) crossed 308 in 50 iters. #176's "use the winning lever from #174" instruction therefore has no concrete value to plug in.

The dependency relationship now flows:

- **If #183 passes** (some (condition, scenario) cell crosses 320 at iter 499): #176 can be unblocked. The winning tuple from #183 (e.g. `normalize_returns=True, value_coef=0.05`) replaces the missing "#174 winner" in #176's CLI invocation. **Action**: re-point #176's `### Dependencies` blocker from #174 → #183 (now resolved), remove `loom:blocked`, and apply the actual winning-lever CLI to its smoke test.
- **If #183 fails** (no cell crosses 320 at iter 499 across either scenario): #176 needs reframing. Its premise — "use policies that actually learn to test whether the conditioner is informative" — collapses, because the ablation budget has not produced learning policies. Likely re-curation paths: (a) test the conditioner on the random-baseline ~308 policies and report what CMI looks like in the no-learning regime as a null distribution; (b) gate #176 on #135 (reward-attribution Python vs Rust core) instead. Either way, #176's current AC is no longer reachable and needs explicit downgrade or replacement.

Either way, **#176 should not be picked up by a Builder until #183 reports**.

### Related

- #174 — Phase 2 ablation; this issue is the long-horizon follow-up (PR #182 ships #174's results).
- #176 — Now gated on this issue (see Dependency note).
- #145 — Phase 1 diagnostic: PPO doesn't learn.
- #151 — Phase 1 plots.
- #159 — return-normalization implementation (PR #169, commit 376366dc).
- #164 — `value_coef` + `entropy_coef` CLI plumbing (commit 807e7fcb).
- #135 — reward-attribution Python vs Rust core; the leading candidate for "next experiment" if this one fails.
- #168 — state-summary CMI conditioner (PR #168, commit 79c04ada); landed but waits on #176, which waits on this issue.

**Comment by rjwalters** [comment]
## Builder status — blocked on remote execution (Path B)

Scripts prepared on `feature/issue-183` (commit 2424bc7c, pushed). Local smoke test passed: 1 cell × 10 iters × 256 steps produced `metrics.json` in the expected directory layout (`runs/p3_183_smoke/L1_norm/default/lambda_0e0/seed_42/`).

**Why blocked:** the SSH hosts named in `CLAUDE.md` (`rwalters-sandbox-1`, `rwalters-sandbox-2`) do not resolve from this builder's environment (`Could not resolve hostname`). Per `CLAUDE.md` "❌ NEVER Run Locally" the 30-cell sweep at 500 iters × 2048 rollout cannot run on the Mac Studio. Surfacing for a human with remote SSH access.

### Files added (review-ready)
- `experiments/p3_specialization/run_183_phase3.sh` — sweep launcher with env-var sharding (`CONDITIONS`, `SCENARIOS`, `SEEDS`, `NUM_ITERATIONS`, `ROLLOUT_STEPS`, `OUTPUT_SUBDIR`, `SKIP_EXISTING`).
- `experiments/p3_specialization/analyze_183.py` — aggregation + plots; Phase 2 (#174) trajectories overlaid on default-scenario panel for visual continuity.

### Recommended remote invocation

On `rwalters-sandbox-2` (or an equivalent compute host):

```bash
ssh rwalters-sandbox-2
cd ~/bucket-brigade
git fetch origin
git checkout feature/issue-183
uv sync --extra rl

# Launch the full sweep in tmux (full 30 cells, ~5-6 cell-hours single-threaded;
# parallelize across shells if the host supports it):
tmux new -d -s phase3 'bash experiments/p3_specialization/run_183_phase3.sh 2>&1 | tee /tmp/phase3.log'

# Or shard across 3 shells (one per condition) to parallelize:
tmux new -d -s phase3_L1   'CONDITIONS=L1_norm bash experiments/p3_specialization/run_183_phase3.sh 2>&1 | tee /tmp/phase3_L1.log'
tmux new -d -s phase3_L1L2 'CONDITIONS=L1L2    bash experiments/p3_specialization/run_183_phase3.sh 2>&1 | tee /tmp/phase3_L1L2.log'
tmux new -d -s phase3_L1L3 'CONDITIONS=L1L3    bash experiments/p3_specialization/run_183_phase3.sh 2>&1 | tee /tmp/phase3_L1L3.log'
```

### After the sweep completes

The Builder will resume on receipt of:
1. `experiments/p3_specialization/runs/p3_183_phase3/` rsync'd back to local (or committed remotely + pulled), and
2. Confirmation that all 30 cells produced `metrics.json`.

The Builder will then run `analyze_183.py` locally (post-hoc plotting is cheap CPU), write the research notebook at `research_notebook/2026-05-15_p3_plateau_phase3.md`, and finalize the PR.

Applying `loom:blocked` per the task spec until a human can execute the sweep on a reachable host.

**Comment by rjwalters** [comment]
Closing as completed. The 500-iteration long-horizon test was run as part of the Phase 2 PPO-tuning ladder; per the project's working notes (`feedback_p3_research.md`, `project_ppo_failure_mode.md`), "the PPO-tuning ladder (#145, #174, #183) is exhausted" — no L1-anchored config closed the learnability gap above the random baseline. Research direction has since pivoted to env-side and CTDE-class interventions, all shipped in the 2026-05-17/18 session and now tracked in #343.
